File: go/Trapdoorless DualRegev/AME.go
package main

import (
	"crypto/rand"
	"fmt"
	"log"
	"math"
	"math/big"
	mrand "math/rand"
	"runtime"
	"sync"
	"time"
)

type BigIntMatrix [][]*big.Int

type Parameters struct {
	q     *big.Int
	p     *big.Int
	m     int
	n     int
	alpha float64
	sigma float64
}

type PublicKey struct {
	par Parameters
	B   BigIntMatrix
}

var rng = mrand.New(mrand.NewSource(time.Now().UnixNano()))

const lambda = 64

// Generates a random matrix of size (rows x cols)
// with entries uniformly sampled from Z_q (integers modulo q).
func sampleMatrix(rows, cols int, q *big.Int) BigIntMatrix {
	matrix := make(BigIntMatrix, rows)
	for i := 0; i < rows; i++ {
		matrix[i] = make([]*big.Int, cols)
		for j := 0; j < cols; j++ {
			// Sample uniformly at random in [0, q-1]
			matrix[i][j] = new(big.Int).Rand(rng, q)
		}
	}
	return matrix
}

// Generates a matrix of size (rows x cols) where each entry
// is sampled from a discrete Gaussian (normal) distribution
// with mean 0 and given standard deviation (stddev).
func sampleError(rows, cols int, stddev float64, mod *big.Int) BigIntMatrix {
	matrix := make(BigIntMatrix, rows)

	for i := 0; i < rows; i++ {
		matrix[i] = make([]*big.Int, cols)
		for j := 0; j < cols; j++ {
			// Draw a sample from N(0, stddev^2)
			x := rng.NormFloat64() * stddev

			// Round to nearest integer
			intVal := int64(math.Round(x))

			// Store as big.Int
			matrix[i][j] = new(big.Int).SetInt64(intVal)
		}
	}
	return matrix
}

// SampleMatrixP generates an m × n matrix where each entry is from P = {-1:1/4, 0:1/2, +1:1/4}
func SampleMatrixP(m, n int) BigIntMatrix {
	matrix := make([][]*big.Int, m)
	for i := 0; i < m; i++ {
		row := make([]*big.Int, n)
		for j := 0; j < n; j++ {
			// Sample from {0, 1, 2, 3} and map to {-1, 0, 1}
			randVal, err := rand.Int(rand.Reader, big.NewInt(4))
			if err != nil {
				log.Fatalf("crand.Int error: %v", err)
			}
			switch randVal.Int64() {
			case 0, 1:
				row[j] = big.NewInt(0)
			case 2:
				row[j] = big.NewInt(-1)
			case 3:
				row[j] = big.NewInt(1)
			}
		}
		matrix[i] = row
	}
	return matrix
}

// Multiplies two matrices (a * b) with entries in Z_mod.
// using parallelism to speed up computation.
func multiplyMatricesParallel(a, b BigIntMatrix, mod *big.Int) BigIntMatrix {
	rows := len(a)     // number of rows in 'a'
	cols := len(b[0])  // number of columns in 'b'
	inner := len(a[0]) // shared inner dimensions (a.cols == b.rows)

	// Allocate result matrix (rows x cols), initialized with zeros
	result := make(BigIntMatrix, rows)
	for i := range result {
		result[i] = make([]*big.Int, cols)
		for j := range result[i] {
			result[i][j] = new(big.Int) // each entry starts as 0
		}
	}

	// Determine number of workers = number of CPU cores
	numWorkers := runtime.NumCPU()
	runtime.GOMAXPROCS(numWorkers)

	// Split the work into batches of rows per worker
	batchSize := (rows + numWorkers - 1) / numWorkers

	var wg sync.WaitGroup
	for w := 0; w < numWorkers; w++ {
		start := w * batchSize
		end := start + batchSize
		if end > rows {
			end = rows
		}
		if start >= rows {
			break
		}

		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()

			tmp := new(big.Int) // temporary for multplication
			sum := new(big.Int) // accumulator for each cell

			// Multiply the corresponding row (from 'a') and column (from 'b')
			for i := start; i < end; i++ {
				for j := 0; j < cols; j++ {
					sum.SetInt64(0) // reset accumulator
					for k := 0; k < inner; k++ {
						tmp.Mul(a[i][k], b[k][j]) // a[i][j] * b[k][j]
						sum.Add(sum, tmp)         // accumulate
					}
					result[i][j].Mod(sum, mod) // reduce modulo 'mod'
				}
			}
		}(start, end)
	}

	wg.Wait() // wait for all goroutines to finish
	return result
}

func Transpose(matrix BigIntMatrix) BigIntMatrix {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return nil
	}

	rows := len(matrix)
	cols := len(matrix[0])

	// Create transposed matrix
	transposed := make(BigIntMatrix, cols)
	for i := range transposed {
		transposed[i] = make([]*big.Int, rows)
	}

	// Fill transposed matrix
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			transposed[j][i] = new(big.Int).Set(matrix[i][j])
		}
	}

	return transposed
}

// Multiplies every entry of matrix 'a' by 'constant'
// and reduces the result modulo 'mod'
func multiplyMatrixByConstant(a BigIntMatrix, constant, mod *big.Int) BigIntMatrix {
	rows, cols := len(a), len(a[0])

	// Allocate result matrix with same dimensions
	result := make(BigIntMatrix, rows)

	// Temporary variable to hold intermediate products
	tmp := new(big.Int)
	for i := 0; i < rows; i++ {
		result[i] = make([]*big.Int, cols)
		for j := 0; j < cols; j++ {
			// Multiply entry by the constant
			tmp.Mul(a[i][j], constant)

			// Reduce modulo 'mod' and store in result
			result[i][j] = new(big.Int).Mod(tmp, mod)
		}
	}
	return result
}

// Adds two matrices 'a' and 'b' element-wise modulo mod
func addMatrices(a, b BigIntMatrix, mod *big.Int) BigIntMatrix {
	rows, cols := len(a), len(a[0])

	// Allocate result matrix
	result := make(BigIntMatrix, rows)

	for i := 0; i < rows; i++ {
		result[i] = make([]*big.Int, cols)
		for j := 0; j < cols; j++ {
			// Add entries a[i][j] + b[i][j]
			sum := new(big.Int).Add(a[i][j], b[i][j])

			// Reduce modulo 'mod' and store in result
			result[i][j] = sum.Mod(sum, mod)
		}
	}
	return result
}

// appendRows stacks two matrices vertically:
// A: r1 x c, B: r2 x c → result: (r1+r2) x c
func appendRows(A, B BigIntMatrix) BigIntMatrix {
	if len(A) == 0 {
		return B
	}
	if len(B) == 0 {
		return A
	}
	colsA := len(A[0])
	colsB := len(B[0])
	if colsA != colsB {
		panic(fmt.Sprintf("appendRows: mismatched column sizes (%d vs %d)", colsA, colsB))
	}

	rows := len(A) + len(B)
	cols := colsA
	result := make(BigIntMatrix, rows)
	for i := 0; i < len(A); i++ {
		result[i] = make([]*big.Int, cols)
		for j := 0; j < cols; j++ {
			result[i][j] = new(big.Int).Set(A[i][j])
		}
	}
	for i := 0; i < len(B); i++ {
		result[len(A)+i] = make([]*big.Int, cols)
		for j := 0; j < cols; j++ {
			result[len(A)+i][j] = new(big.Int).Set(B[i][j])
		}
	}
	return result
}

// RoundDiv returns ⌊a/b⌉ (round to nearest integer).
func RoundDiv(a, b *big.Int) *big.Int {
	q, r := new(big.Int), new(big.Int)
	q.QuoRem(a, b, r) // q = a/b (truncated), r = remainder

	// If remainder >= b/2, round up
	half := new(big.Int).Div(b, big.NewInt(2))
	if r.Cmp(half) >= 0 {
		q.Add(q, big.NewInt(1))
	}
	return q
}

func genParameters() Parameters {
	mrand.Seed(time.Now().UnixNano())
	p := big.NewInt(128)
	q := new(big.Int).Lsh(big.NewInt(1), 18)

	// Sample dimensions m, n in poly(lambda)
	m := lambda*lambda + mrand.Intn(11)
	n := lambda + mrand.Intn(11)

	// Sample error rate alpha
	alpha := 1.0 / (2.0 * float64(q.Int64()))
	sigma := 1.0

	return Parameters{p: p, q: q, m: m, n: n, alpha: alpha, sigma: sigma}
}

func kGen() (BigIntMatrix, PublicKey) {
	par := genParameters()

	// Sample matrix A from (n x m)
	A := sampleMatrix(par.n, par.m, par.q)
	s := SampleMatrixP(par.m, 1)

	// Compute s^T * A^T mod q
	AT := Transpose(A)
	sT := Transpose(s)
	As := multiplyMatricesParallel(sT, AT, par.q)

	// Public key: vertically stack A^T and As
	B := appendRows(AT, As)

	pk := PublicKey{par: par, B: B}
	return s, pk
}

func enc(pk PublicKey, mu BigIntMatrix) BigIntMatrix {
	par := pk.par
	B := pk.B
	delta := new(big.Int).Div(par.q, par.p)

	// Sample uniform vector r
	r := sampleMatrix(par.n, 1, par.q)

	// Sample error vectors e0 and e1
	e0 := sampleError(par.m, 1, par.sigma, par.q)
	e1 := sampleError(1, 1, par.sigma, par.q)

	// Compute B*r
	Br := multiplyMatricesParallel(B, r, par.q)

	// Vertically stack e0 and e1
	e := appendRows(e0, e1)

	// Vertically stack 0^m and mu
	zeros := make(BigIntMatrix, par.m)
	for i := 0; i < par.m; i++ {
		zeros[i] = make([]*big.Int, 1) // 1 column
		zeros[i][0] = big.NewInt(0)    // Initialize with 0
	}
	mu_stack := appendRows(zeros, mu)
	delta_mustack := multiplyMatrixByConstant(mu_stack, delta, par.q)

	// Compute c = Br + e + delta*mu_stack
	tmp := addMatrices(Br, e, par.q)
	c := addMatrices(tmp, delta_mustack, par.q)

	return c
}

func dec(par Parameters, sk, ct BigIntMatrix) *big.Int {
	delta := new(big.Int).Div(par.q, par.p)

	// Construct sk_neg = [-sk^T | 1]
	skT := Transpose(sk)
	sk_neg := make(BigIntMatrix, len(skT))
	for i := range skT {
		sk_neg[i] = make([]*big.Int, len(skT[i])+1)
		for j := range skT[i] {
			tmp := new(big.Int).Neg(skT[i][j])
			sk_neg[i][j] = tmp.Mod(tmp, par.q)
		}
		sk_neg[i][len(skT[i])] = big.NewInt(1)
	}

	// Compute nu = [-sk | 1] * ct
	nu := multiplyMatricesParallel(sk_neg, ct, par.q)
	nuVal := new(big.Int).Set(nu[0][0])
	qdiv := RoundDiv(nuVal, delta)
	mu := new(big.Int).Mod(qdiv, par.p)

	return mu
}

func main() {
	iterations := 1000
	regularSuccess := 0

	for i := 0; i < iterations; i++ {
		fmt.Println("Iteration round:", i+1)
		sk, pk := kGen()
		par := pk.par
		mu := sampleMatrix(1, 1, par.p)
		ct := enc(pk, mu)
		dm := dec(par, sk, ct)

		if mu[0][0].String() == dm.String() {
			regularSuccess++
		}
	}

	fmt.Println(regularSuccess, "/", iterations, "regular decryptions succeeded!")

}

File: python/Trapdoorless DualRegev/AME.py
import math
import numpy as np
import random

# Security Parameter
lam = 16

# ----------------- Helper Sampling Functions -----------------

def sample_uniform_matrix(rows, cols, q):
    """
    Sample a matrix with uniform random integers in [0, q-1].
    Used for public matrix generation in GSW.
    """
    return np.random.randint(0, q-1, size=(rows, cols), dtype=int)


def sample_error_matrix(rows, cols, stddev, q):
    """
    Sample a small "error" matrix from a discrete Gaussian (rounded normal distribution).
    Used to introduce noise in GSW ciphertexts.
    """
    return np.round(np.random.normal(loc=0, scale=stddev, size=(rows, cols))).astype(int)


def sample_matrix_p(m, n):
    """
    Sample an m x n matrix from P ? {-1, 0, 1} with probabilites:
    -1 with 1/4, 0 with 1/2, +1 with 1/4
    Used for secret key generation.
    """
    # Sample integers from {0,1,2,3}
    r = np.random.randint(0, 4, size=(m, n))

    # Map values: {0,1 -> 0; 2 -> -1; 3 -> 1}
    matrix = np.where(r < 2, 0, np.where( r == 2, -1, 1))
    return matrix


# ----------------- Gadget Functions -----------------

def gadget_matrix(n, k, q):
    g = 2 ** np.arange(k)
    G = np.kron(np.eye(n, dtype=int), g) % q
    return G


def calculateSMatrix(k, l, q):
    # Extract bits from LSB to MSB
    q_bin = bin(q)[2:]  # remove the '0b' prefix

    q_bits = [int(b) for b in q_bin]


    # Build Sk
    Sk = np.zeros((k,k), dtype=int)
    for i in range(k):
        Sk[i,i] = 2
        if i > 0:
            Sk[i,i-1] = -1
        Sk[i,-1] = q_bits[i]   # last column

    # Build full S
    I = np.eye(l, dtype=int)
    S = np.kron(I, Sk)
    return S


# ----------------- Parameter Generation -----------------

def gen_parameters():

    # Sample modulus p, q poly(lambda)
    p = 128
    q = 2**18

    # Sample dimensions m, n = poly(lambda)
    m = lam**2 + random.randint(0, 10)
    n = lam + random.randint(0, 10)

    # Compute k
    k_upper = m**2 - math.ceil(p*lam*m/2)
    k_bound = m - (n*math.log2(q) + 2*lam)
    k_max = min(k_upper, k_bound)
    k = random.randint(1, max(1, int(k_max)))

    # Sample error rate alpha in (0,1)
    alpha = 1/(2*q)
    # Standard deviation sigma > alpha*q
    sigma = 1.5
    return q, p, m, n, alpha, sigma


# ----------------- Key Generation -----------------

def kgen():
    """
    Key generation for Dual-GSW.
    Outputs secret key sk and public key pk.
    """
    par = gen_parameters()
    q, p, m, n, alpha, sigma = par

    A = sample_uniform_matrix(n, m, q)
    s = sample_matrix_p(m, 1)

    # Compute s^T * A^T mod q
    As = np.matmul(s.T, A.T) % q

    # Public key: vertically stack A^T and As
    B = np.vstack([A.T, As])

    pk = (par, B)
    sk = s
    return sk, pk


# ----------------- Anamorphic Key Generation -----------------

def agen():
    # Generate parameters
    par = gen_parameters()
    q, p, m, n, alpha, sigma = par

    # Compute k (your heuristic; keep as-is)
    k = math.ceil((m/2) - math.ceil(math.sqrt((lam*m)/2)))
    limit = n*math.log2(q)+2*lam
    if m-k <= limit:
        k = math.ceil(k/2)

    print("m=", m, " k=", k)

    # Sample s <- P^m and define J = { j : s_j = 0}
    s = sample_matrix_p(m, 1)
    J =[i for i in range(m) if s[i, 0] == 0]

    # If |J| < k, abort (retry a few times)
    max_tries = 10
    tries = 0
    while len(J) < k and tries < max_tries:
        s = sample_matrix_p(m, 1)
        J = [i for i in range(m) if s[i, 0] == 0]
        tries += 1
    if len(J) < k:
        return None, None, None, None

    # Choose I subset uniformly of size k from J
    I = sorted(random.sample(J, k))

    # Sample A_bar and T
    A_bar = sample_uniform_matrix(m - k, n, q)   # shape (m-k, n)
    T = sample_matrix_p(m-k, k)                  # shape (m-k, k)

    # Compute last k rows: T^T * A_bar => shape (k, n)
    TT_Abar = (T.T @ A_bar) % q

    # Build A' of shape (m x n): stack A_bar then TT_Abar
    A_prime = np.vstack([A_bar, TT_Abar]) % q  # (m, n)

    # Permute A' rows to form A according to I and complement
    full_indices = list(range(m))
    complement = [idx for idx in full_indices if idx not in I]
    if len(complement) != (m - k):
        raise AssertionError("complement size mismatch")

    A = np.zeros((m, n), dtype=np.int64)

    # CORRECTED: place A_prime[j,:] in complement[j]
    for j in range(m - k):
        idx = complement[j]
        A[idx, :] = A_prime[j, :] % q

    for j in range(k):
        idx = I[j]
        A[idx, :] = A_prime[m - k + j, :] % q

    # Compute As = s^T * A  (1 x n)
    As = (s.T @ A) % q

    # Public key B = [ A ; As ]  -> shape (m+1, n)
    B = np.vstack([A, As]) % q

    # Build T' := [ -T^T | I_k ]  (k x m)
    neg_TT = (-T.T) % q
    I_k = np.eye(k, dtype=np.int64) % q
    Tprime = np.hstack([neg_TT, I_k]) % q  # shape (k, m)

    # Reorder columns of Tprime to form eT
    eT = np.zeros((k, m), dtype=np.int64)
    for j in range(m - k):
        col_dst = complement[j]
        eT[:, col_dst] = Tprime[:, j] % q
    for j in range(k):
        col_dst = I[j]
        eT[:, col_dst] = Tprime[:, (m - k) + j] % q

    # Sanity checks
    lhs = (Tprime @ A_prime) % q
    if not np.all(lhs == 0):
        print("Warning: Tprime @ A_prime.T != 0 (mod q)")

    te_at = (eT @ A) % q
    if not np.all(te_at == 0):
        print("Warning: eT @ A.T != 0 (mod q)")

    apk = (par, B)
    ask = s
    dk = (k, I)
    tk = eT

    return ask, apk, dk, tk


# ----------------- Encryption -----------------

def enc(pk, mu):
    """
    Encrypt a plaintext mu using Dual Regev.
    pk: public key
    mu: plaintext integer in [0, p-1]
    """
    par = pk[0]
    B = pk[1]
    q, p, m, n, alpha, sigma = par
    delta = math.ceil(q/p)

    # Sample uniformly vector r
    r = sample_uniform_matrix(n, 1, q)

    # Sample error vectors e0 and e1
    e0 = sample_error_matrix(m, 1, sigma, q)
    e1 = sample_error_matrix(1, 1, sigma, q)

    # Compute B*r
    Br = np.matmul(B, r)

    # Vertically stack e0 and e1
    e = np.vstack([e0, e1])

    # Vertically stack O^m and mu
    zeros = np.zeros((m,1), dtype=int)
    mu_stack = np.vstack([zeros, np.array([mu])])
    delta_mustack = delta*mu_stack

    # Compute c = Br + e + delta*mustack
    c = (Br + e + delta_mustack) % q
    return c


# ----------------- Anamorphic Encryption -----------------

def aenc(apk, dk, mu, mu_hat):
    par, B = apk
    q, p, m, n, alpha, sigma = par

    k, I = dk

    # Delta = ceil(q/p)
    delta = int(math.ceil(q / p))

    # Sample uniform vector r
    r = sample_uniform_matrix(n, 1, q)

    # Sample error vector e0 and e1
    e0 = sample_error_matrix(m, 1, sigma, q)
    e1 = sample_error_matrix(1, 1, sigma, q)

    # Compute f_mu_hat (mx1) with mu_hat entries placed at indices I
    f = np.zeros((m,1), dtype=np.int64)
    for j, idx in enumerate(I):
        # idx is 0-based index in [0...m-1]
        f[idx, 0] = int(mu_hat[j])

    # Compute Br = B * r mod q
    Br = np.matmul(B, r) % q

    # Stack e = [e0; e1]
    e = np.vstack([e0, e1])

    # Stack delta*[f; mu]
    mu_col = np.array([[int(mu)]], dtype=np.int64)
    f_and_mu = np.vstack([f, mu_col])
    delta_term = (delta * f_and_mu) % q

    # Compute c = Br + e + delta*f_and_mu
    c = (Br + e + delta_term) % q
    return c


# ----------------- Decryption -----------------

def dec(par, sk, ct):
    """
    Decrypt a ciphertext ct using Dual Regev.
    Returns plaintext integer mu.
    """
    q, p, m, n, alpha, sigma = par
    delta_float = q/p
    delta = math.ceil(q/p)
    print(delta_float)
    print(delta)

    # Construct sk_neg = [-sk^T | 1]
    sk_neg = np.concatenate([-sk.T, np.array([[1]])], axis=1)

    # Compute nu = [-sk^T | 1] * ct
    nu = np.matmul(sk_neg, ct) % q
    mu = int(np.round(nu.item() / delta)) % p
    return mu


# ----------------- Anamorphic Decryption -----------------

def adec(par, dk, tk, ask, act):
    k, I = dk
    q, p, m, n, alpha, sigma = par
    delta = int(math.ceil(q/p))

    # Build [Te | 0] (k x (m+1))
    zero_col = np.zeros((k, 1), dtype=np.int64)
    Te_ext = np.hstack([tk, zero_col])

    nu_hat = np.matmul(Te_ext, act) % q

    mu_hat = np.round(nu_hat / delta).astype(np.int64) % p

    return mu_hat


# ----------------- Testing -----------------

# Number of trials
iterations = 1000
regular_success = 0
anamorphic_success = 0

sk, pk = kgen()
par = pk[0]
q, p, m, n, alpha, sigma = par

for i in range(iterations):
    mu = random.randint(0, p-1)
    ct = enc(pk, mu)
    dm = dec(par, sk, ct)

    if mu == dm:
        regular_success += 1
    else:
        print("Original message: ",mu, " and decrypted message: ", dm)

ask, apk, dk, tk = agen()
par = apk[0]
q, p, m, n, alpha, sigma = par

for i in range(iterations):
    mu = random.randint(0, p-1)
    mu_hat = sample_uniform_matrix(dk[0], 1, p)
    act = aenc(apk, dk, mu, mu_hat)

    adm = adec(par, dk, tk, ask, act)

    if np.array_equal(mu_hat, adm):
        anamorphic_success += 1
    else:
        print("Anamorphic message: ", mu_hat.T, " and decrypted message: ", adm.T)

print(f"{regular_success}/{iterations} regular decryptions succeeded!")
print(f"{anamorphic_success}/{iterations} anamorphic decryptions succeeded!")